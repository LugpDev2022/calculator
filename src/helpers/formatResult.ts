import convertToScientificNotation from './convertToScientificNotation';

const formatResult = (result: number, error?: string): string => {
  if (error) {
    return error;
  }

  const absResult = Math.abs(result);

  if (absResult >= 10000000000000 && Number.isFinite(result)) {
    return convertToScientificNotation(result);
  }

  return result.toLocaleString('en-US');
};

export default formatResult;
