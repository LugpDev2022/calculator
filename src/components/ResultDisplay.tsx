import formatResult from '../helpers/formatResult';

interface Props {
  error?: string;
  result: number;
}

const ResultDisplay: React.FC<Props> = ({ result, error = '' }) => {
  const displayedResult = formatResult(result, error);

  return (
    <span
      className='
        block
        font-bold
        text-2xl
        text-right
        w-full
        sm:text-4xl
      '
    >
      {displayedResult}
    </span>
  );
};

export default ResultDisplay;
