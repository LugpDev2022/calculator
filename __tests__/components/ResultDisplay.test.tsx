import React from 'react';
import { vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import formatResult from '../../src/helpers/formatResult';
import ResultDisplay from '../../src/components/ResultDisplay';

vi.mock('../../src/helpers/formatResult', () => ({
  default: vi.fn(),
}));

describe('Tests on <ResultDisplay />', () => {
  beforeEach(() => vi.clearAllMocks());

  test('should match snapshot', () => {
    (formatResult as jest.Mock).mockReturnValue('1057');

    const { container } = render(<ResultDisplay result={1057} />);

    expect(container).toMatchSnapshot();
  });

  test('should display the result', () => {
    (formatResult as jest.Mock).mockReturnValue('1057');

    render(<ResultDisplay result={1057} />);

    expect(screen.getByText('1057')).toBeTruthy();
  });

  test('should call formatResult() without error', () => {
    (formatResult as jest.Mock).mockReturnValue('1057');

    const result = 1057;

    render(<ResultDisplay result={result} />);

    expect(formatResult).toHaveBeenCalledWith(result, '');
  });

  test('should call formatResult() with the given error', () => {
    (formatResult as jest.Mock).mockReturnValue('1057');

    const result = 1057;
    const error = 'Invalid Operation';

    render(<ResultDisplay result={result} error={error} />);

    expect(formatResult).toHaveBeenCalledWith(result, error);
  });
});
