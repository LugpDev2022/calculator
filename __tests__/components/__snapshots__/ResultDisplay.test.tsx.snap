// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <ResultDisplay /> > should match snapshot 1`] = `
<div>
  <span
    class=" block font-bold text-2xl text-right w-full sm:text-4xl "
  >
    1057
  </span>
</div>
`;
